### PR TITLE
[WIP] FileUploads improvement

### PR DESCRIPTION
**Pull request by sopvop**
This is a wip branch for review and bikeshedding. do not merge.
Implements ideas discussed in #235

I've implemented foldMultipart and handleMultiparts in terms of it.

```haskell
foldMultipart ::
       (MonadSnap m) =>
       UploadPolicy        -- ^ global upload policy
    -> PartFold a          -- ^ part processor
    -> a                   -- ^ seed accumulator
    -> m a
```

All tests seem to pass, but I have additional tests I want to add, though now it properly checks the number of uploaded form inputs I think at some point tests were passing even though there were no checks in place.

### CODE REVIEW

**Comment by sopvop** [comment]
Number of form limits test seem to be working ok.
Now on high level api. I propose something simpler than handleFileUploads
```haskell
data FileUploadPolicy = FileUploadPolicy
    { maxFileSize :: Int64
    , maxNumFiles :: Int
    }

handleFormData :: MonadSnap m 
         => UploadPolicy 
         -> FileUploadPolicy 
         -> (PartInfo -> InputStream ByteString -> a)
         -> m [(Maybe ByteString, a)]

storeAsLazyByteString :: PartInfo -> InputStream ByteString -> LB.ByteString

withTemporaryStore :: FilePath -> (PartInfo -> InputStream ByteString) -> LB.ByteString

myUploadsHandler  = do
   withTemporaryStore "/var/tmp" $ \store -> do
       files <- handleFormData defaultUploadPolicy defaultFileUploadPolicy store
       processMyFiles files

mySmallFilesHandler = do
     let filePolicy = setMaximumFileSize 1024 defaultFileUploadPolicy
     filesaslbs <- handleFormData defaultUploadPolicy filePolicy storeAsLazyByteString
```

Another API thing which came up with files was people requesting that form data was not inserted into rqParams and instead was as a separate dict.

Alternatively we can instead change foldMultipart to not post process form inputs into rqParams and return them as list, like so
```haskell
data FormParam = FormParam !ByteString !ByteString
foldMultipart ::
       (MonadSnap m) =>
       UploadPolicy        -- ^ global upload policy
    -> PartFold a          -- ^ part processor
    -> a                   -- ^ seed accumulator
    -> m ([FormParam], a)
```
But old handleUploads api would still store them as rqParams.

Thoughts?

**Comment by mightybyte** [comment]
This is looking nice.  I think I would prefer returning the list explicitly to setting up a separate dict.

@gregorycollins?

**Comment by sopvop** [comment]
Implemented high level interface, which is used like this:
```haskell
myFormHandler = withTemporaryStore "/var/tmp" "upload.tmp" $ \store -> do
    (params, files) <- handleFormUploads defaultUploadPolicy defaultFileUploadPolicy store
    processFiles files


smallFileUploads = do
    let filePlicy = setMaximumFileSize (512*1024) defaultFileUploadPolicy
    (params, files) <- handleFormUploads defaultUploadPolicy filePolicy storeAsLazyByteString
    processByteStrings files

```

## High level API
```haskell
type FormFile a = (Maybe ByteString, a)
type FormParam = (ByteString, ByteString)

handleFormUploads ::
       (MonadSnap m) =>
       UploadPolicy                   -- ^ general upload policy
    -> FileUploadPolicy               -- ^ Upload policy for files
    -> (PartInfo -> InputStream ByteString -> IO a)
                                      -- ^ A file storage function
    -> m ([FormParam], [FormFile a])
```
## storage helpers

```haskell
storeAsLazyByteString :: PartInfo -> InputStream ByteString -> IO LB.ByteString

withTemporaryStore ::
    MonadSnap m
    => FilePath -- ^ temporary directory
    -> String   -- ^ file name pattern
    -> ((PartInfo -> InputStream ByteString -> IO FilePath) -> m a)
      -- ^ Action taking store function
    -> m a
```
## config
```haskell
data FileUploadPolicy = FileUploadPolicy { ... }

setMaximumFileSize :: Int64 -> FileUploadPolicy -> FileUploadPolicy
setMaximumNumberOfFiles :: Int -> FileUploadPolicy -> FileUploadPolicy
setSkipFilesWithoutNames:: Int64 -> FileUploadPolicy -> FileUploadPolicy
setStoreFilesWithoutNames :: Int64 -> FileUploadPolicy -> FileUploadPolicy
```

### Why "skipFilesWithoutNames" is needed
The setSkipFilesWithoutNames is useful to use with html5 forms.
According to html5 spec, form inputs with type file if not set are
encoded as files without name, with empty body and mime type of
application/octet-stream. Default FileUploadPolicy allows empty files
with zero length, it should fit this use-case. Note that older browsers
may omit such fields from encoded form-data, so this behaviour is
desireable.


## defaultFileUploadPolicy
maxFileSize is 1mb
maxNumberOfFiles is 10, but probably 1 would be a safer behaviour
Files without names are skipped with maximum size 0 (zero).

## Tests
I think everything is reasonable tested, but If you think of any corner cases I would gladly add tests for it.

I would like to get as much bike-shedding as possible before I do changes to module documentation.

**Comment by mightybyte** [comment]
I haven't had time to look at this in-depth, but I like what I'm seeing.  Hopefully @gregorycollins will be able to take a look at it soon.

File: src/Snap/Internal/Util/FileUploads.hs [hunk 1]
```diff
@@ -172,15 +192,87 @@ handleFileUploads tmpdir uploadPolicy partPolicy partHandler =
 
 
 ------------------------------------------------------------------------------
+type FormFile a = (Maybe ByteString, a)
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Prefer named fields here -- first parameter should probably be strict.

File: src/Snap/Internal/Util/FileUploads.hs [hunk 2]
```diff
@@ -172,15 +192,87 @@ handleFileUploads tmpdir uploadPolicy partPolicy partHandler =
...
+       UploadPolicy                   -- ^ general upload policy
+    -> FileUploadPolicy               -- ^ Upload policy for files
+    -> (PartInfo -> InputStream ByteString -> IO a)
+                                      -- ^ A file storage function
+    -> m ([FormParam], [FormFile a])
+handleFormUploads uploadPolicy filePolicy partHandler = do
+    (params, !st) <- foldMultipart uploadPolicy go (UploadState 0 id)
+    return (params, uploadedFiles st [])
+  where
+    go partInfo stream st = do
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Put bangs on 'partInfo' and 'st'?

File: src/Snap/Internal/Util/FileUploads.hs [hunk 3]
```diff
@@ -198,19 +290,20 @@ type PartProcessor a = PartInfo -> InputStream ByteString -> IO a
 -- /Exceptions/
 --
 -- If the given 'UploadPolicy' stipulates that you wish form inputs to be
--- placed in the 'rqParams' parameter map (using 'setProcessFormInputs'), and
--- a form input exceeds the maximum allowable size, this function will throw a
--- 'PolicyViolationException'.
+-- processed (using 'setProcessFormInputs'), and form a form input exceeds
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Typo in the comment here

File: src/Snap/Internal/Util/FileUploads.hs [hunk 4]
```diff
@@ -198,19 +290,20 @@ type PartProcessor a = PartInfo -> InputStream ByteString -> IO a
...
+-- the maximum allowable size or form exceeds maximum number of inputs, this
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: ..or the form exceeds its maximum number of inputs

File: src/Snap/Internal/Util/FileUploads.hs [hunk 5]
```diff
@@ -528,6 +638,69 @@ setUploadTimeout s u = u { uploadTimeout = s }
 
 
 ------------------------------------------------------------------------------
+
+data EmptyFileNamePolicy = EmptyFileNameStore Int64
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Just wondering why this isn't just this under the hood:

```haskell
-- better names of course ;)
data EmptyFileNamePolicy = EFN { skip :: !Bool, maxSize :: !Int64 }
```

Seems like some of the logic is simpler if you do this?

File: src/Snap/Internal/Util/FileUploads.hs [hunk 6]
```diff
@@ -198,19 +290,20 @@ type PartProcessor a = PartInfo -> InputStream ByteString -> IO a
...
--- 'PolicyViolationException'.
+-- processed (using 'setProcessFormInputs'), and form a form input exceeds
+-- the maximum allowable size or form exceeds maximum number of inputs, this
+-- function will throw a 'PolicyViolationException'.
 --
 -- If an uploaded part contains MIME headers longer than a fixed internal
 -- threshold (currently 32KB), this function will throw a 'BadPartException'.
 --
-handleMultipart ::
+foldMultipart ::
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Can you also please add a "Since" entry for this function? I.e.

```haskell
-- /Since: 1.0.2.0/
```

You'll also need to bump the minor version number to 1.0.2.0 in the .cabal.

File: src/Snap/Internal/Util/FileUploads.hs [hunk 7]
```diff
@@ -245,29 +337,47 @@ handleMultipart uploadPolicy origPartHandler = do
...
+    -> m [a]
+handleMultipart uploadPolicy origPartHandler = do
+    (captures, files) <- foldMultipart uploadPolicy partFold id
+    procCaptures captures
+    return $! files []
+
+  where
+    partFold info input acc = do
+      x <- origPartHandler info input
+      return $ acc . ([x]++)
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Prefer `(x:)` to `([x]++)` here?

File: src/Snap/Internal/Util/FileUploads.hs [hunk 8]
```diff
@@ -639,7 +900,7 @@ internalHandleMultipart !boundary clientHandler !stream = go
             throwIO $ BadPartException "headers exceeded maximum size"
 
     --------------------------------------------------------------------------
-    goPart !str = do
+    goPart !str !acc = do
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: "state" is a better name than acc here?

File: src/Snap/Internal/Util/FileUploads.hs [hunk 9]
```diff
@@ -649,30 +910,45 @@ internalHandleMultipart !boundary clientHandler !stream = go
...
+                                     }
+
+    throwTooMuchVars =
+        throwIO . PolicyViolationException
+        $ T.concat [ "number of form inputs exceeded maximum of "
+                   , T.pack $ show maxFormVars ]
 
     --------------------------------------------------------------------------
-    processMixed !fieldName !str !mixedBoundary = do
+    processMixed !fieldName !str !acc !mixedBoundary= do
```
Comment (gregorycollins) [inline src/Snap/Internal/Util/FileUploads.hs]: Spurious diff

**Comment by sopvop** [comment]
Sorry for long delay. But at least I've figured out possible problems with api.

Reworked this api quite a bit. Folded all the empty name policy options into `FileUploadPolicy`.
Now it is can be configured with:
```haskell
setMaximumFileSize :: Int64 -> FileUploadPolicy -> FileUploadPolicy
setMaximumNumberOfFiles :: Int -> FileUploadPolicy -> FileUploadPolicy
setSkipFilesWithoutNames :: Bool -> FileUploadPolicy -> FileUploadPolicy
setMaximumSkippedFileSize :: Int64 -> FileUploadPolicy -> FileUploadPolicy
```

If `setSkipFilesWithoutNames` is `False`, then the `setMaximumSkippedFileSize` is ignored and `setMaximumFileSize` is used as for normal files.


`handleFormUploads` now returns a specialized pair 
```haskell
data FormFile a = FormFile
    { formFileName  :: !ByteString
         -- ^ Name of a field
    , formFileValue :: a
         -- ^ Result of storing file
    } deriving (Eq, Ord, Show)
```

I guess I should also update module docs.

**Comment by gregorycollins** [comment]
Great! We should plan to deprecate the old functions if the new ones offer a strict superset of the functionality.
